src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TcpTransportCreationHandler.cs

```diff
@@ -0,0 +1,193 @@
+        {
+            // This handler cannot process if the protocol does not contain TCP
+            if (parameters.DataSource.ResolvedProtocol is not (DataSource.Protocol.Admin or DataSource.Protocol.TCP or DataSource.Protocol.None))
+            {
+                return null;
+            }
+
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            IPAddress[] ipAddresses = isAsync
+                ? await Dns.GetHostAddressesAsync(parameters.DataSource.ServerName, ct).ConfigureAwait(false)
+                : Dns.GetHostAddresses(parameters.DataSource.ServerName);
+            if (ipAddresses is null || ipAddresses.Length == 0)
+            {
+                throw new SocketException((int)SocketError.HostNotFound);
+            }
+
+            // If there is an IP version preference, apply it
+            switch (parameters.ConnectionString.IPAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionComparer.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionComparer.InstanceV6);
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketOpenExceptions = new List<Exception>();
+
+            int portToUse = parameters.DataSource.ResolvedPort < 0
+                ? parameters.DataSource.Port
+                : parameters.DataSource.ResolvedPort;
+            var ipEndPoint = new IPEndPoint(IPAddress.None, portToUse); // Allocate once
+            foreach (IPAddress ipAddress in ipAddresses)
+            {
+                ipEndPoint.Address = ipAddress;
+                try
+                {
+                    socket = await OpenSocket(ipEndPoint, isAsync, ct).ConfigureAwait(false);
+                    break;
+                }
+                catch(Exception e)
+                {
+                    socketOpenExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                // If there are any socket exceptions in the collected exceptions, throw the first
+                // one. If there are not, collect all exceptions and throw them as an aggregate.
+                foreach (Exception exception in socketOpenExceptions)
+                {
+                    if (exception is SocketException)
+                    {
+                        throw exception;
+                    }
+                }
+
+                throw new AggregateException(socketOpenExceptions);
+            }
+
+            // Create the stream for the socket
+            return new NetworkStream(socket);
```

Comment:
Does disposing NetworkStream dispose socket also? If not, then can you add it to the context, so that it can be disposed, when the connection is closed ?